### PR TITLE
[docs][kubernetes] GPU user guide

### PR DESCRIPTION
**Pull request by DmitriGekhtman**
## Why are these changes needed?

This PR 
- adds a page of guidance on GPU deployment with Ray/K8s. This page is a modified and slightly expanded version of the existing page  https://docs.ray.io/en/latest/cluster/kubernetes-gpu.html
- moves managed K8s service intro links to their own page 

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by DmitriGekhtman** [comment]
TODO, after getting more reference/intro docs in, is to write an e2e guide on a sample workload involving GPUs

File: doc/source/cluster/cluster_under_construction/ray-clusters-on-kubernetes/user-guides/gpu.rst [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+
+The KubeRay operator advertises container GPU resource limits to
+the Ray scheduler and the Ray autoscaler. In particular, the Ray container's
+`ray start` entrypoint will be automatically configured with the appropriate `--num-gpus` option.
+
+* Learn more about Ray's :ref:`gpu-support`.
+
+GPU workload scheduling
+~~~~~~~~~~~~~~~~~~~~~~~
+After a Ray pod with access to GPU is deployed, it will
```
Comment (pcmoritz) [inline doc/source/cluster/cluster_under_construction/ray-clusters-on-kubernetes/user-guides/gpu.rst:67]: Can we also add some info about what is required (if anything) to make the ML libraries use GPUs? E.g. TensorFlow or Pytorch. Maybe link to the section in the documentation that talks about this (I hope there is one).

File: doc/source/cluster/cluster_under_construction/ray-clusters-on-kubernetes/user-guides/gpu.rst [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+    Finally, make sure your group or pool of GPU Kubernetes nodes are configured to autoscale.
+
+GPUs and Ray
+____________
+
+The KubeRay operator advertises container GPU resource limits to
+the Ray scheduler and the Ray autoscaler. In particular, the Ray container's
+`ray start` entrypoint will be automatically configured with the appropriate `--num-gpus` option.
+
+* Learn more about Ray's :ref:`gpu-support`.
```
Comment (pcmoritz) [inline doc/source/cluster/cluster_under_construction/ray-clusters-on-kubernetes/user-guides/gpu.rst:63]: This is currently rendered as an enumeration with a single entry, should we just attach it to the previous paragraph?

File: doc/source/cluster/cluster_under_construction/ray-clusters-on-kubernetes/user-guides/gpu.rst (hunk 1 above)
Comment (pcmoritz) [reply doc/source/cluster/cluster_under_construction/ray-clusters-on-kubernetes/user-guides/gpu.rst:67]: (This is probably what most people will be interested in doing)

File: doc/source/cluster/cluster_under_construction/ray-clusters-on-kubernetes/user-guides/gpu.rst [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+be able to execute tasks and actors annotated with gpu requests.
+For example, the decorator `@ray.remote(num_gpus=1)` annotates a task or actor
+requiring 1 GPU.
+
+
+GPU autoscaling
+~~~~~~~~~~~~~~~
+The Ray autoscaler is aware of each Ray worker group's GPU capacity.
+Say we have a RayCluster configured as in the config snippet above:
+- There is a worker group of Ray pods with 1 unit of GPU capacity each
```
Comment (pcmoritz) [inline doc/source/cluster/cluster_under_construction/ray-clusters-on-kubernetes/user-guides/gpu.rst:77]: This is not interpreted as an enumeration by rst currently (you might need *)

**Review by pcmoritz — APPROVED** [review]

File: doc/source/cluster/cluster_under_construction/ray-clusters-on-kubernetes/user-guides/gpu.rst (hunk 2 above)
Comment (DmitriGekhtman) [reply doc/source/cluster/cluster_under_construction/ray-clusters-on-kubernetes/user-guides/gpu.rst:63]: Yeah -- my bad.

File: doc/source/cluster/cluster_under_construction/ray-clusters-on-kubernetes/user-guides/gpu.rst (hunk 1 above)
Comment (DmitriGekhtman) [reply doc/source/cluster/cluster_under_construction/ray-clusters-on-kubernetes/user-guides/gpu.rst:67]: I think the `rayproject/ray-ml:gpu` are (or should be) packaged with everything required for the Ray ML libraries covered in the Ray docs.

I will make this assertion.
@richardliaw does that sound reasonable?

File: doc/source/cluster/cluster_under_construction/ray-clusters-on-kubernetes/user-guides/gpu.rst [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+   ...
+   containers:
+   - name: ray-node
+     image: rayproject/ray:nightly-gpu
+     ...
+
+Node selectors and node labels
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+To ensure Ray pods are bound to Kubernetes nodes satisfying specific
+conditions (such as presence of GPU hardware), you may wish to use
```
Comment (cadedaniel) [inline doc/source/cluster/cluster_under_construction/ray-clusters-on-kubernetes/user-guides/gpu.rst:180]: nit: There's a missing article: `s/such as presence of GPU hardware/such as the presence of GPU hardware/g`

File: doc/source/cluster/cluster_under_construction/ray-clusters-on-kubernetes/user-guides/gpu.rst [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+            ...
+
+Each of the Ray pods in the group can be scheduled on an AWS `p2.xlarge` instance (1 GPU, 4vCPU, 61Gi RAM).
+
+.. tip::
+
+    GPU instances are expensive -- consider setting up autoscaling for your GPU Ray workers,
+    as demonstrated with the `minReplicas:0` and `maxReplicas:5` settings above.
+    To enable autoscaling, remember also to set `enableInTreeAutoscaling:True` in your RayCluster's `spec`
+    Finally, make sure your group or pool of GPU Kubernetes nodes are configured to autoscale.
```
Comment (cadedaniel) [inline doc/source/cluster/cluster_under_construction/ray-clusters-on-kubernetes/user-guides/gpu.rst:54]: Should there be a link or instructions for this?

File: doc/source/cluster/cluster_under_construction/ray-clusters-on-kubernetes/user-guides/gpu.rst [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+Requesting GPUs
+~~~~~~~~~~~~~~~
+You can also make a :ref:`direct request to the autoscaler<ref-autoscaler-sdk-request-resources>` to scale up GPU resources.
+
+.. code-block:: python
+
+    import ray
+
+    ray.init()
+    ray.autoscaler.sdk.request_resources(bundles=[{"GPU": 1} * 2])
```
Comment (cadedaniel) [inline doc/source/cluster/cluster_under_construction/ray-clusters-on-kubernetes/user-guides/gpu.rst:114]: `[{"GPU": 1} * 2]` doesn't work in Python3.7. Maybe you meant `[{"GPU": 1}] * 2`?

What's the plan for adding CI for these examples? I don't know how it works, but I do see the other doc sections use `literalinclude` to include a selection from a file inside the `source/<topic>/doc_code` directory.

File: doc/source/cluster/cluster_under_construction/ray-clusters-on-kubernetes/user-guides/gpu.rst (hunk 5 above)
Comment (DmitriGekhtman) [reply doc/source/cluster/cluster_under_construction/ray-clusters-on-kubernetes/user-guides/gpu.rst:54]: It's cloud-specific but usually straightforward (check the autoscaling box)
Will add a link to the page introduced in this PR on managed K8s services.

File: doc/source/cluster/cluster_under_construction/ray-clusters-on-kubernetes/user-guides/gpu.rst (hunk 6 above)
Comment (DmitriGekhtman) [reply doc/source/cluster/cluster_under_construction/ray-clusters-on-kubernetes/user-guides/gpu.rst:114]: Thanks for pointing the correction!

About testing -- For now, I will do the test manually before merging.
We already have CI that covers the functionality.

I will however open an issue to track covering this exact example in the CI -- will have to come back to that later, since we're on a tight timeline at the moment.

File: doc/source/cluster/cluster_under_construction/ray-clusters-on-kubernetes/user-guides/gpu.rst (hunk 4 above)
Comment (DmitriGekhtman) [reply doc/source/cluster/cluster_under_construction/ray-clusters-on-kubernetes/user-guides/gpu.rst:180]: Thanks, will add the article :)

I wonder what the actual syntax rules here are
`Specific metaphysical conditions (such as existence)` but
`Specific conditions of your software system (such as _the_ presence of GPU hardware)`

File: doc/source/cluster/cluster_under_construction/ray-clusters-on-kubernetes/user-guides/gpu.rst (hunk 6 above)
Comment (DmitriGekhtman) [reply doc/source/cluster/cluster_under_construction/ray-clusters-on-kubernetes/user-guides/gpu.rst]: https://github.com/ray-project/ray/issues/27388

File: doc/source/cluster/cluster_under_construction/ray-clusters-on-kubernetes/user-guides/gpu.rst (hunk 5 above)
Comment (DmitriGekhtman) [reply doc/source/cluster/cluster_under_construction/ray-clusters-on-kubernetes/user-guides/gpu.rst:65]: Linked.

File: doc/source/cluster/cluster_under_construction/ray-clusters-on-kubernetes/user-guides/gpu.rst (hunk 4 above)
Comment (DmitriGekhtman) [reply doc/source/cluster/cluster_under_construction/ray-clusters-on-kubernetes/user-guides/gpu.rst]: Article added.

File: doc/source/cluster/cluster_under_construction/ray-clusters-on-kubernetes/user-guides/gpu.rst (hunk 3 above)
Comment (DmitriGekhtman) [reply doc/source/cluster/cluster_under_construction/ray-clusters-on-kubernetes/user-guides/gpu.rst:91]: Fixed by adding a space.
rst whitespace syntax is gross

File: doc/source/cluster/cluster_under_construction/ray-clusters-on-kubernetes/user-guides/gpu.rst (hunk 2 above)
Comment (DmitriGekhtman) [reply doc/source/cluster/cluster_under_construction/ray-clusters-on-kubernetes/user-guides/gpu.rst]: Fixed.

File: doc/source/cluster/cluster_under_construction/ray-clusters-on-kubernetes/user-guides/gpu.rst (hunk 1 above)
Comment (DmitriGekhtman) [reply doc/source/cluster/cluster_under_construction/ray-clusters-on-kubernetes/user-guides/gpu.rst:80]: Added a bit of exposition.

**Comment by pcmoritz** [comment]
As a follow up, we should also flesh out the `Managed Kubernetes services` section a bit (or disband it and make it part of another section) -- at the moment it looks a little sad :)

We can address it in a follow up PR and think about how it fits into the rest of the K8s docs.

File: doc/source/cluster/cluster_under_construction/ray-clusters-on-kubernetes/user-guides/gpu.rst (hunk 6 above)
Comment (DmitriGekhtman) [reply doc/source/cluster/cluster_under_construction/ray-clusters-on-kubernetes/user-guides/gpu.rst]: Fixed, tracked, tested manually.

**Comment by DmitriGekhtman** [comment]
> As a follow up, we should also flesh out the `Managed Kubernetes services` section a bit (or disband it and make it part of another section) -- at the moment it looks a little sad :)
> 
> We can address it in a follow up PR and think about how it fits into the rest of the K8s docs.

sgtm, probably it will be best rolled into something getting-started-ish

**Comment by DmitriGekhtman** [comment]
removed 2.0.0 pick label, as docs PRs will require some special handling

**Comment by pcmoritz** [comment]
@DmitriGekhtman Can you elaborate on that? Is the process described somewhere? So far I thought all changes that need to be picked need the label :)

**Comment by DmitriGekhtman** [comment]
> @DmitriGekhtman Can you elaborate on that? Is the process described somewhere? So far I thought all changes that need to be picked need the label :)


Deferring to @scv119 for this logistics questions.

**Comment by cadedaniel** [comment]
> > @DmitriGekhtman Can you elaborate on that? Is the process described somewhere? So far I thought all changes that need to be picked need the label :)
> 
> Deferring to @scv119 for this logistics questions.

I can answer! We’re moving the Ray Clusters docs over to the new structure in several PRs. We’re going to create a single cherry-pick PR for all of them at the end so it’s easier to manage.
